protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SetAnnotationCommandParser.java

```diff
@@ -0,0 +1,72 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.decode.parser;
+
+import java.util.List;
+
+import org.apache.james.imap.api.ImapCommand;
+import org.apache.james.imap.api.ImapConstants;
+import org.apache.james.imap.api.ImapMessage;
+import org.apache.james.imap.api.display.HumanReadableText;
+import org.apache.james.imap.api.process.ImapSession;
+import org.apache.james.imap.decode.ImapRequestLineReader;
+import org.apache.james.imap.decode.base.AbstractImapCommandParser;
+import org.apache.james.imap.message.request.SetAnnotationRequest;
+import org.apache.james.mailbox.model.MailboxAnnotation;
+import org.apache.james.protocols.imap.DecodingException;
+
+import com.google.common.collect.Lists;
+
+public class SetAnnotationCommandParser extends AbstractImapCommandParser {
+
+    public SetAnnotationCommandParser() {
+        super(ImapCommand.authenticatedStateCommand(ImapConstants.SETANNOTATION_COMMAND_NAME));
+    }
+
+    @Override
+    protected ImapMessage decode(ImapCommand command, ImapRequestLineReader request, String tag, ImapSession session)
+            throws DecodingException {
+        final String mailboxName = request.mailbox();
+        char next = request.nextWordChar();
+        List<MailboxAnnotation> listMailboxAnnotations = Lists.newArrayList();
+        if (next == '(') {
+            request.consumeChar('(');
+            do {
+                try {
+                    String key = request.atom();
+                    String value = request.nstring();
+                    if (null == value) {
+                        listMailboxAnnotations.add(MailboxAnnotation.nil(key));
+                    } else {
+                        listMailboxAnnotations.add(MailboxAnnotation.newInstance(key, value));
+                    }
+                } catch (IllegalArgumentException e) {
+                    throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "The key is not valid: " + e.getMessage());
+                }
+    
+                next = request.nextWordChar();
```

Comment:
Variable next is useless : You can directly call request.nextWorldChar without storing the value.